### PR TITLE
Add CloudFormation template for VPN + MFA auth layer (single PR)

### PR DESCRIPTION
**Pull request by chmreid**
This PR smooshes all 8 pull requests for the VPN server (#51, #54, #55, #56) and VPN auth layer (#69, #70, #72, #73) into one giant pull request. It is identical though (all commits in those PRs are in this PR). The small PRs should make it easier to review individual parts of the proposed changes, and this PR will make it easier to discuss in a single place and merge the changes in once they're ready.

### CODE REVIEW

File: openvpn.yml.template [hunk 1]
```diff
@@ -0,0 +1,855 @@
...
+    us-west-1:  # Northern California
+      "ubuntu": "ami-0f56279347d2fa43e"  # Ubuntu Server Linux AMI 2018.04
+    us-west-2:  # Oregon
+      "ubuntu": "ami-003634241a8fcdec0"  # Ubuntu Server Linux AMI 2018.04
+
+Resources:
+  dockstoreVpnVPC:
+    Type: AWS::EC2::VPC
+    Properties:
+      CidrBlock: 10.0.0.0/22
```
Comment (coverbeck) [inline openvpn.yml.template:137]: 1. Do these CIDR blocks have the potential of clashing with a CIDR block from existing VPC?
2. Would it make sense to move the VPC creation into core? I guess it's a separate VPC, so maybe not, but on the other hand, you have the ability to deal with CIDR overlaps better.

File: openvpn.yml.template [hunk 2]
```diff
@@ -0,0 +1,855 @@
...
+    Properties:
+      GroupDescription: Security group for OpenVPN Server
+      VpcId: !Ref dockstoreVpnVPC
+      SecurityGroupIngress:
+        - IpProtocol: !Ref OpenVPNProtocol
+          FromPort: !Ref OpenVPNPort
+          ToPort: !Ref OpenVPNPort
+          CidrIp: 0.0.0.0/0
+        - IpProtocol: tcp
+          FromPort: 22
```
Comment (coverbeck) [inline openvpn.yml.template:215]: Do we need/want this? We don't have ssh enabled for our other EC2 instances.

Comment (chmreid) [reply openvpn.yml.template:220]: Oh, that's my bad, I was using SSH to connect to the instance. I will remove that.

File: openvpn.yml.template (hunk 1 above)
Comment (chmreid) [reply openvpn.yml.template:142]: You raise a great point - this CIDR block does clash with the existing VPCs. However, that will only matter if resources are on _both_ VPCs. It's theoretically possible but for our setup I don't think we'd ever do that.

I would prefer to keep the VPN VPC in the openvpn.yml file for the sake of separation of concerns - they've stayed separate so far and it would be nice if we could keep it that way. I ended up adopting the same technique used in core.yml (define the Cidr block prefix using a variable), and bumped the VPC CIDR block prefix to 10.2 instead of 10.0, to help alleviate potential confusion. I also created a variable for the VPN CIDR block prefix (the IPs handed out for the VPN network itself, currently 10.200) so that should also help with clarity.

Thanks for pointing this out!

Comment (david4096) [reply openvpn.yml.template:142]: Just to clarify, since these are on different VPCs the CIDR blocks can be the same and the core VPC will only see traffic from the internet gateway IP in the VPN VPC. Changing the second byte on the VPN network is a way to make it more clear to clients what network they are on if they are using resources within the VPN network, but there isn't a risk of collision, right?

File: openvpn.yml.template [hunk 3]
```diff
@@ -0,0 +1,861 @@
...
+
+  CertificateEmail:
+    Description: The value to set for the certificate's Email field.
+    Type: String
+    Default: dockstore@dockstore.org
+
+  OpenVPNPort:
+    Description: OpenVPN port number.
+    Type: Number
+    Default: 1194
```
Comment (david4096) [inline openvpn.yml.template:112]: I think it might be worth changing the default to make it more difficult to discover the service?

File: openvpn.yml.template [hunk 4]
```diff
@@ -0,0 +1,861 @@
...
+              command: !Sub |
+                aws s3 cp /var/log/cfn-init.log s3://${dockstoreVpnS3Bucket}/log/cfn-init.log
+
+Outputs:
+  bucketName:
+    Description: UUID for the S3 bucket
+    Value: !Ref dockstoreVpnS3Bucket
+  eip:
+    Description: Instance EIP
+    Value: !Ref dockstoreVpnEIP
```
Comment (david4096) [inline openvpn.yml.template:861]: This suggests some future PR that allows the LoadBalancer to be configured automatically to accept this address.

File: openvpn.yml.template (hunk 1 above)
Comment (chmreid) [reply openvpn.yml.template:142]: Correct, there is no risk of collision even if they have the same subnet IP, because they don't share a gateway.

File: openvpn.yml.template (hunk 3 above)
Comment (chmreid) [reply openvpn.yml.template:112]: My initial reaction is to say it isn't worth it. Security through obscurity is not going to stop dedicated attackers - if we have buggy software it'll be broken no matter what port we're on, and port scanners can reveal more info on what services are running. So, it only stops amateur attackers/non-targeted attacks.

The other reason is, you need a server certificate to attempt a login username/password so random internet people can't attempt drive-by brute force attacks. If it were something like SSH with passwords enabled, where you just need a username and password to log in, then I would say it is worth changing port numbers.

Comment (chmreid) [reply openvpn.yml.template:112]: On the other hand if random internet people can attempt to log in, we should probably change the port number. I will look into whether this is a risk and follow up with more information.

File: openvpn.yml.template (hunk 1 above)
Comment (chmreid) [reply openvpn.yml.template]: Updated template to use CIDR block variables, marking this resolved

File: openvpn.yml.template (hunk 2 above)
Comment (chmreid) [reply openvpn.yml.template]: Removed the open ssh port, marking as resolved

File: openvpn.yml.template [hunk 5]
```diff
@@ -0,0 +1,906 @@
...
+    Default: udp
+    AllowedValues:
+    - udp
+    - tcp
+    Description: OpenVPN protocol.
+
+Mappings:
+  RegionMap:
+    us-east-1:  # Northern Virginia
+      "ubuntu": "ami-085925f297f89fce1"  # Ubuntu Server Linux AMI 2018.04
```
Comment (denis-yuen) [inline openvpn.yml.template:137]: Comment might want to include version number for debugging in the future

File: openvpn.yml.template [hunk 6]
```diff
@@ -0,0 +1,906 @@
...
+          Ebs:
+            VolumeSize: 20
+      Tags:
+      - Key: Name
+        Value: OpenVPN Server
+      UserData:
+          "Fn::Base64":
+            !Sub |
+              #!/bin/bash -xe
+              wget https://s3.amazonaws.com/amazoncloudwatch-agent/ubuntu/amd64/latest/amazon-cloudwatch-agent.deb -O /tmp/amazon-cloudwatch-agent.deb
```
Comment (denis-yuen) [inline openvpn.yml.template:318]: Not a change for this PR, but hopefully the CDK makes this kind of global requirement easier

File: openvpn.yml.template [hunk 7]
```diff
@@ -0,0 +1,1128 @@
...
+                # Allow incoming SSH sessions, new or established
+                /sbin/iptables -A INPUT -p tcp --dport 22 -m conntrack --ctstate NEW,ESTABLISHED -j ACCEPT
+                # Allow incoming SSH traffic, if part of established conversation
+                /sbin/iptables -A INPUT -p tcp --sport 22 -m conntrck --ctstate ESTABLISHED -j ACCEPT
+                ########### PING ##############
+                # Allow ping requests
+                /sbin/iptables -A INPUT -p icmp --icmp-type echo-request -j ACCEPT
+                ########### VPN ###############
+                # Allow incoming VPN sessions destined for 1194, new or established
+                /sbin/iptables -A INPUT -p ${OpenVPNProtocol} --dport 1194 -m conntrack --ctstate NEW,ESTABLISHED -j ACCEPT
```
Comment (david4096) [inline openvpn.yml.template:706]: Since the port is configurable this might break if its changed from the default.

File: openvpn.yml.template [hunk 8]
```diff
@@ -0,0 +1,1128 @@
...
+        configure_iptables:
+          packages:
+            apt:
+              netfilter-persistent: []
+          files:
+            /root/iptables_openvpn_dns.sh:
+              content: !Sub |
+                #!/bin/bash
+                # Set default policies
+                /sbin/iptables -P INPUT DROP
```
Comment (david4096) [inline openvpn.yml.template:684]: Do you need to set up iptables on an instance? I was under the impression that type of operation is done using Security Groups. There are cases where you might want to modify iptables and the internal firewall that I might not be comprehending here though.

File: openvpn.yml.template (hunk 7 above)
Comment (chmreid) [reply openvpn.yml.template:706]: Good catch! Thank you.

File: openvpn.yml.template [hunk 9]
```diff
@@ -0,0 +1,1128 @@
...
+                # Flush and clear everything
+                /sbin/iptables -t nat -F
+                /sbin/iptables -t mangle -F
+                /sbin/iptables -F
+                /sbin/iptables -X
+                ########### INCOMING ##########
+                # Allow any established connection to come in or out
+                iptables -A INPUT -m state --state RELATED,ESTABLISHED -j ACCEPT
+                iptables -A OUTPUT -m state --state RELATED,ESTABLISHED -j ACCEPT
+                ########### SSH ###############
```
Comment (chmreid) [inline openvpn.yml.template:696]: Forgot I had this bit to allow SSH in the iptables rule, for testing. Will remove that.

File: openvpn.yml.template (hunk 8 above)
Comment (chmreid) [reply openvpn.yml.template:684]: iptables gives finer-grained control over packet flow in the system. For example, the Security Groups can allow incoming/outgoing traffic on a particular port, but iptables can allow or deny traffic on a particular port based on whether it is a new conversation or part of an existing conversation. It is also necessary to masquerade external DNS requests and allow the VPN server to act as a router. The AWS Security Group is a more blunt instrument with one purpose, while iptables is a more flexible instrument serving multiple purposes.

File: openvpn.yml.template [hunk 10]
```diff
@@ -0,0 +1,1128 @@
...
+                # Allow incoming SSH sessions, new or established
+                /sbin/iptables -A INPUT -p tcp --dport 22 -m conntrack --ctstate NEW,ESTABLISHED -j ACCEPT
+                # Allow incoming SSH traffic, if part of established conversation
+                /sbin/iptables -A INPUT -p tcp --sport 22 -m conntrck --ctstate ESTABLISHED -j ACCEPT
+                ########### PING ##############
```
Comment (chmreid) [inline openvpn.yml.template:701]: Should limit accepted ping packets to those received through the VPN tunnel only (so that machines on the VPN can still ping the server, but it cannot be pinged via its external IP). Also need to adjust the security group to implement the same rule.

File: openvpn.yml.template (hunk 8 above)
Comment (chmreid) [reply openvpn.yml.template:684]: (Also in case you're wondering about the 127.0.0.1 and 127.0.0.53 addresses in the DNS iptables rules, we're actually running _two_ DNS servers on the OpenVPN box. Linux runs its own built-in DNS resolver at 127.0.0.53, and we are running dnsmasq at 127.0.0.1. This is part of the reason we need the complicated iptables setup.)

File: openvpn.yml.template [hunk 11]
```diff
@@ -0,0 +1,1128 @@
...
+                # Allow ping requests
+                /sbin/iptables -A INPUT -p icmp --icmp-type echo-request -j ACCEPT
+                ########### VPN ###############
+                # Allow incoming VPN sessions destined for 1194, new or established
+                /sbin/iptables -A INPUT -p ${OpenVPNProtocol} --dport 1194 -m conntrack --ctstate NEW,ESTABLISHED -j ACCEPT
+                # Allow incoming VPN traffic coming from 1194, part of established conversation
+                /sbin/iptables -A INPUT -p ${OpenVPNProtocol} --sport 1194 -m conntrck --ctstate ESTABLISHED -j ACCEPT
+                ########### DNS ###############
+                PROTOCOLS="tcp udp"
```
Comment (chmreid) [inline openvpn.yml.template:710]: This needs a comment explaining that 127.0.0.53 is the built-in Ubuntu DNS server (resolved), while 127.0.0.1 is the custom DNS server for the VPN (dnsmasq), and that these iptables rules are needed so that DNS requests from VPN clients pass through the dnsmasq DNS server at 127.0.0.1 first, then to the system DNS at 127.0.0.53.

File: openvpn.yml.template [hunk 12]
```diff
@@ -0,0 +1,1128 @@
...
+                domain=dockstore.vpn
+                # add the local domain to the end of hostnames
+                expand-hosts
+                # listen on these interfaces
+                listen-address=127.0.0.1
+                listen-address=${DockstoreVpnCidrPrefixVpn}.0.1
+                bind-interfaces
+                # define special addresses to resolve
+                # TODO: figure out how this should work with multiple stages
+                address=/gateway.dockstore.net/10.200.0.1
```
Comment (chmreid) [inline openvpn.yml.template:789]: `s/10.200/${DockstoreVpnCidrPrefixVpn}/`

File: openvpn.yml.template (hunk 9 above)
Comment (chmreid) [reply openvpn.yml.template:696]: Removed in commit f22bb50

File: openvpn.yml.template (hunk 10 above)
Comment (chmreid) [reply openvpn.yml.template:701]: Added in 5fe8eac
- iptables ping rules
- security group ping rules

File: openvpn.yml.template (hunk 12 above)
Comment (chmreid) [reply openvpn.yml.template:789]: Fixed in commit 70107f4

File: openvpn.yml.template (hunk 10 above)
Comment (chmreid) [reply openvpn.yml.template:701]: Added in commit 5fe8eac
- iptables ping rules
- security group ping rules

File: openvpn.yml.template (hunk 9 above)
Comment (chmreid) [reply openvpn.yml.template:696]: Removed in commit f22bb50

File: openvpn.yml.template (hunk 11 above)
Comment (chmreid) [reply openvpn.yml.template:710]: Fixed in f502c66

File: grafana.yml.template [hunk 13]
```diff
@@ -16,6 +16,10 @@ Parameters:
     Description: The name of the Dockstore stack
     Type: String
     MinLength: 1
+  VpnStack:
```
Comment (coverbeck) [inline grafana.yml.template:19]: This will require changes to dockstore-deploy.sh as well and the template files, but I can tackle that separately.

This also means that the order of deploy of stacks will be:

1. Core, DB and other stacks
2. Dockstore
3. Grafana
4. VPN

The README should reflect this, but I noticed I didn't update it myself when I added the Grafana stack.

File: openvpn.yml.template (hunk 4 above)
Comment (coverbeck) [reply openvpn.yml.template:1229]: Following up on this comment, it's not clear to me how this will work in the overall flow. How will we be able to go to dev.dockstore.net:3000 and get to Grafana, since there is a security group that won't allow this, and I don't see an update to the LB's security group?

Or is the expectation at this point that you navigate directly to the Grafana's public IP?

File: grafana.yml.template (hunk 13 above)
Comment (chmreid) [reply grafana.yml.template:19]: Actually the Grafana stack should be deployed _after_ the OpenVPN stack, since the OpenVPN stack will create Grafana's VPN credentials and export a few variables, and the Grafana stack will import those variables, grab the VPN credentials from the S3 bucket, and set up and join the VPN.

Comment (coverbeck) [reply grafana.yml.template:19]: OK, I had incorrectly assumed it was using the value to import from another stack like the other stack properties. But looking at it more closely, this property isn't used at all, so should this property even be present?

File: openvpn.yml.template (hunk 4 above)
Comment (chmreid) [reply openvpn.yml.template:1229]: Thanks for bringing this up. Here's a little more info:
* The grafana instance will use the OpenVPN server's public IP address to join the VPN and establish an encrypted connection to the OpenVPN server
* The grafana instance will be on the virtual private network at the private IP 10.200.0.5 (that IP is always the same)
* When a VPN user wants to connect to the grafana instance, their traffic will first go to the OpenVPN server, which will already have a connection to the grafana server and will be able to forward the packet on to the grafana server. In other words, users do not connect directly to the grafana instance, they connect to the VPN server, which acts as a router for the network and directs traffic where it needs to go
* Note that means users DO NOT type `dev.dockstore.net:3000` and get a connection to the grafana server (that would require a public DNS record to redirect to a private IP address, which violates DNS rules)
* Instead, users type `grafana.dockstore.vpn:3000`, and a DNS server running on the OpenVPN server will translate `grafana.dockstore.vpn` into `10.200.0.5`, thus redirecting the user's traffic to the grafana instance
* You can use the above approach for ANY stage; if you connect to the dev OpenVPN server, `grafana.dockstore.vpn` will redirect you to the dev grafana server; if you connect to the staging OpenVPN server, `grafana.dockstore.vpn` will redirect you to the staging grafana server; etc.

You are right that the load balancer was not updated to accept traffic on port 1194. Unfortunately, port 1194 is a variable in the OpenVPN stack, so to avoid specifying the port number in two cloudformation stacks, we need to change the deploy order so the OpenVPN stack gets deployed first. That way, it can export the port number variable as an importable variable, and the dockstore stack can import it and know which port to open on the load balancer.

It would definitely make things easier to specify the OpenVPN port number using an SSM parameter instead, to keep the Dockstore stack from relying on the OpenVPN stack. I will go ahead and make that change and re-request your review on the PR once it's been added.

**Comment by chmreid** [comment]
@coverbeck left a response to your comment about how the overall flow looks when connecting to grafana, but it doesn't seem to be showing up, so I'm including it below to make it easier to find.

-----


Thanks for bringing this up. Here's a little more info:

* The grafana instance will use the OpenVPN server's public IP address to join the VPN and establish an encrypted connection to the OpenVPN server
* The grafana instance will be on the virtual private network at the private IP 10.200.0.5 (that IP is always the same)
* When a VPN user wants to connect to the grafana instance, their traffic will first go to the OpenVPN server, which will already have a connection to the grafana server and will be able to forward the packet on to the grafana server. In other words, users do not connect directly to the grafana instance, they connect to the VPN server, which acts as a router for the network and directs traffic where it needs to go
* Note that means users DO NOT type dev.dockstore.net:3000 and get a connection to the grafana server (that would require a public DNS record to redirect to the private VPN IP address of the grafana, which violates DNS rules)
* Instead, users type grafana.dockstore.vpn:3000, and a DNS server running on the OpenVPN server will translate grafana.dockstore.vpn into 10.200.0.5, thus redirecting the user's traffic to the grafana instance
* You can use the above approach for ANY stage; if you connect to the dev OpenVPN server, grafana.dockstore.vpn will redirect you to the dev grafana server; if you connect to the staging OpenVPN server, grafana.dockstore.vpn will redirect you to the staging grafana server; etc.

You are right that the load balancer was not updated to accept traffic on port 1194. Unfortunately, port 1194 is a variable in the OpenVPN stack, so to avoid specifying the port number in two cloudformation stacks, we need to change the deploy order so the OpenVPN stack gets deployed first. That way, it can export the port number variable as an importable variable, and the dockstore stack can import it and know which port to open on the load balancer.

It would definitely make things easier to specify the OpenVPN port number using an SSM parameter instead, to keep the Dockstore stack from relying on the OpenVPN stack. I will go ahead and make that change and re-request your review on the PR once it's been added.

**Comment by coverbeck** [comment]
@chmreid Thanks for clarifying. 

I have probably a naive view of VPNs, and haven't delved into nor understood all the details of your PR, nor ever worked with a VPN server, but what was going on my head was something like this:

* I connect to the VPN
* Ignoring advanced stuff like split tunneling, all my requests to the internets go through the VPN server
* The VPN's IP address is allowlisted on the load balancer, so a request that goes to dev.dockstore.net:3000 goes through to Grafana.

**Comment by chmreid** [comment]
>  all my requests to the internets go through the VPN server

This is not how the VPN server currently operates. There is an OpenVPN configuration option that would make the VPN work this way, but it is disabled. The reason is, we don't want the OpenVPN server to be handling non-Dockstore traffic and DNS requests, as it could potentially violate the privacy of VPN users and could result in significantly more traffic (e.g., if you download a 100 MB file, forgetting you're connected to the VPN, all that traffic passes through the VPN server).

The VPN server does not receive all traffic to a user, but if the user types `grafana.dockstore.vpn`, no DNS server will be able to resolve that address except the VPN DNS server. The user's computer will first ask its default DNS service for the DNS record for `grafana.dockstore.vpn`, which will not be found, and it will go down the list of all DNS servers, until it gets to the VPN's DNS server. That DNS server will be able to resolve `grafana.dockstore.vpn`. Once the user queries the VPN DNS and is able to resolve `grafana.dockstore.vpn` to `10.200.0.5`, it will send traffic to `10.200.0.5` via the gateway `10.200.0.1` (that's the OpenVPN server).

The traffic from the user to the OpenVPN server and from the OpenVPN server to the grafana server is encrypted and inside the VPN tunnel during transit, so when the packet crosses the load balancer security group, it will appear as a packet coming FROM the public IP of the OpenVPN server, going TO the OpenVPN server, on port 1194. That's the load balancer security group rule that will be implemented (only allow incoming port 1194 packets from the public IP of the OpenVPN server).

Once the VPN traffic reaches the grafana instance, then the OpenVPN client running on the grafana instance will decrypt and unwrap the packet, where it will appear as a packet coming from a VPN client and going to 10.200.0.5 port 3000. In other words, (the packet coming from the VPN client going to the Grafana client via the OpenVPN server on port 3000) is wrapped up in (a packet coming from the OpenVPN public IP going to the load balancer public IP on port 1194).

Once the packet reaches the grafana instance at port 1194, it is "unwrapped" (decrypted) on the grafana machine, where the packet becomes a packet  from the VPN client to the grafana VPN IP on port 3000.

**Comment by coverbeck** [comment]
> This is not how the VPN server currently operates.

Yes, I understand now that it doesn't work that way. I was partially responding to, which I should have quoted, to your previous comment "that would require a public DNS record to redirect to the private VPN IP address of the grafana, which violates DNS rules", that I wasn't proposing such a violation, and explained how I thought it was going to work.

I'd still prefer being able to go to dev.dockstore.net for Grafana, but we could optionally tackle that later.

File: grafana.yml.template (hunk 13 above)
Comment (chmreid) [reply grafana.yml.template:19]: This was due to the mess that I made while merging the changes that split out grafana into a separate yml file. I'll go through again and double check my changes.

**Comment by chmreid** [comment]
> I'd still prefer being able to go to dev.dockstore.net for Grafana

This should be possible, by defining a DNS record in `dnsmasq.conf` that makes dev.dockstore.net resolve to the VPN IP address of the Grafana instance. But if the DNS request for dev.dockstore.net goes to a public DNS server first, instead of the VPN's dnsmasq server, then `dev.dockstore.net` will return the Load Balancer's public IP, and traffic will be blocked by the security group when trying to connect to dev.dockstore.net.

Currently the dnsmasq server on the VPN is added to the client machine's list of DNS servers to query via the OpenVPN config file. I'm not sure if the dnsmasq server will be added to the _top_ or the _bottom_ of a client machine's list of DNS servers. If it's added to the top, it's ok, but if it's added to the bottom, we will have to figure out another solution.

**TL;DR:** dev.dockstore.net redirecting to grafana on the VPN should be possible, but will require some investigation to confirm.

**Comment by chmreid** [comment]
One more thing - this change would mean dev.dockstore.net port 3000 would redirect to Grafana instance, but port 80 and 443 would also redirect to Grafana instance (since not going through load balancer), so you wouldn't be able to visit/look at dev _and_ view grafana simultaneously

**Comment by chmreid** [comment]
Switched from lowerCamelCase to UpperCamelCase everywhere in 136b47d

File: openvpn.yml.template [hunk 14]
```diff
@@ -0,0 +1,1225 @@
...
+      PolicyDocument:
+        Version: '2012-10-17'
+        Statement:
+          - Action:
+              - ec2:AssociateAddress
+              - ec2:DescribeInstances
+            Effect: Allow
+            Resource: "*"
+          - Action:
+              - s3:*
```
Comment (david4096) [inline openvpn.yml.template:295]: If I understand correctly this policy is attached to the Dockstore instances role. Is it also assumed by the user? The s3: * tipped me off because I believe we should be able to describe the specific bucket it works with and not give broad S3 privileges perhaps? `!Sub '${VpcName}-vpns3bucket'`

**Comment by chmreid** [comment]
@david4096 thanks for your feedback on this PR. Your comment on the S3 policy raised some issues with how IAM permissions are laid out in this cloudformation template - but yes, you are correct that it is unnecessarily broad in granting the VPN EC2 instance access to all S3 buckets and not just the VPN bucket.

>    this policy is attached to the Dockstore instances role

Only to the Dockstore VPN EC2 instance role - not to all Dockstore instances. (That's probably what you meant, but just to be clear.) The Roles attribute of this IAM policy means it is attached to that role, and that role gets attached to the EC2 instance profile.

**Comment by chmreid** [comment]
Found a few lingering issues:
* Policy for S3 buckets on VPN instance can be more restrictive, and only allow access to/from specific bucket
* Policy for S3 buckets on Grafana instance is not explicit (and now that I look at it, I'm not sure how that worked in testing, since I didn't have an S3 access policy in testing either)
* The load balancer security group for Grafana traffic can be more restrictive, and only allow traffic to/from the OpenVPN instance's public IP address (thought I had done this earlier, but turns out I had only restricted to the VPN port)

**Comment by chmreid** [comment]
Re the lambda function:

> are we going to put the lambda with the other lambdas and/or rewrite in js? https://github.com/dockstore/lambda

Lambda has already been rewritten in JS, but moving it into the lambda repo would complicate this PR too much. For now I want to keep it a simple inline script, and tackle separating it later.

> would it be more, or less support to move the scripts to separate files and include them? authoring scripts in a cloudformation template doesn't seem like a great experience

I think I mentioned this somewhere already, but if a lambda function's code is not in the CloudFormation template, then you must provide the path to a zip file in a bucket that has the lambda function's layers bundled up, meaning the deploy step requires additional scripting and buckets to automate, which we may or may not have ready. So, we can address that step in a later PR, especially as the CSP report filtering function requires more customization.
